File: 프로그래머스/5/49190. 방의 개수/방의 개수.py
from collections import defaultdict

def solution(arrows):
    directions=[[-1,0],[-1,1],[0,1],[1,1],[1,0],[1,-1],[0,-1],[-1,-1]]
    r=c=0
    answer = 0
    nodes = defaultdict(bool)
    edges =  defaultdict(bool)
    nodes[f"{r}_{c}"]=True
    
    for dir in arrows:
        for _ in range(2):
            nr, nc = r+directions[dir][0],  c+directions[dir][1]
            if(( nodes[f"{nr}_{nc}"] and not( edges[f"{r}_{c}_{nr}_{nc}"]))):
                answer+=1
            
            nodes[f"{nr}_{nc}"]=True
            edges[f"{r}_{c}_{nr}_{nc}"]=True
            edges[f"{nr}_{nc}_{r}_{c}"]=True
            r, c=nr, nc
            

    return answer

File: 프로그래머스/5/49190. 방의 개수/방의 개수.js
function solution(arrows) {
    
    const directions=[
        [-1,0],// 0
        [-1,1],// 1
        [0,1], // 2
        [1,1], // 3
        [1,0], // 4
        [1,-1],// 5
        [0,-1], // 6
        [-1,-1] // 7
    ]
    const nodes = new Map()
    const routes = new Map()
    let [r,c] = [0,0] 
    nodes.set(`${r}_${c}`,true)
    let answer = 0
    
    for(let dir of arrows){
        const [dr, dc] = directions[dir]
        // console.log(dir)
        
        for(let i=0;i<2;i++){
            const [nr, nc] = [r+dr, c+dc]
            const currentKey = `${r}_${c}`
            const nextKey = `${nr}_${nc}`
            const currentRoute = `${currentKey}_${nextKey}`
            const reverseRoute = `${nextKey}_${currentKey}`
            const hasNode = nodes.get(nextKey)
            const hasRoute = routes.get(currentRoute)
            // console.log(hasNode, hasRoute)
            if(hasNode & !hasRoute){
                answer+=1
            }
            
            
            routes.set(currentRoute, true)
            routes.set(reverseRoute,true)
            nodes.set(nextKey, true)
            
            r = nr
            c = nc
            
        }
    }
    return answer;
}